python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,259 @@
+        data["last-partition-id"] = max(spec["field-id"] for spec in data["partition-spec"])
+        data["default-sort-order-id"] = 0
+        return data
+
+    @root_validator()
+    def migrate_schema(cls, data: Dict[str, Any]):
+        # Migrate schemas
+        schema = data["schema_"]
+        schemas = data["schemas"]
+        if all([schema != other_schema for other_schema in schemas]):
+            data["schemas"].append(schema)
+        data["current_schema_id"] = schema.schema_id
+        return data
+
+    @root_validator()
+    def migrate_partition_spec(cls, data: Dict[str, Any]):
+        # This is going to be much nicer as soon as partition-spec is also migrated to pydantic
+        if partition_spec := data.get("partition_spec"):
+            data["partition_specs"] = [{**spec, "spec-id": INITIAL_SPEC_ID + idx} for idx, spec in enumerate(partition_spec)]
+            data["default_spec_id"] = INITIAL_SPEC_ID
+            data["last_partition_id"] = max(spec["spec-id"] for spec in data["partition_specs"])
+        return data
+
+    table_uuid: Optional[UUID] = Field(alias="table-uuid")
+    """A UUID that identifies the table, generated when the table is created.
+    Implementations must throw an exception if a table’s UUID does not match
+    the expected UUID after refreshing metadata."""
+
+    format_version: Literal[1] = Field(alias="format-version")
+    """An integer version number for the format. Currently, this can be 1 or 2
+    based on the spec. Implementations must throw an exception if a table’s
+    version is higher than the supported version."""
+
+    schema_: Schema = Field(alias="schema")
+    """The table’s current schema. (Deprecated: use schemas and
+    current-schema-id instead)"""
+
+    partition_spec: List[Dict[str, Any]] = Field(alias="partition-spec")
+    """The table’s current partition spec, stored as only fields.
+    Note that this is used by writers to partition data, but is
+    not used when reading because reads use the specs stored in
+    manifest files. (Deprecated: use partition-specs and default-spec-id
+    instead)"""
+
+
+class TableMetadataV2(TableMetadataCommonFields, IcebergBaseModel):
+    @root_validator(skip_on_failure=True)
+    def check_if_schema_is_found(cls, data: Dict[str, Any]):
+        current_schema_id = data["current_schema_id"]
+
+        for schema in data["schemas"]:
+            if schema.schema_id == current_schema_id:
+                return data
+
+        raise ValueError(f"current-schema-id {current_schema_id} can't be found in the schemas")
+
+    @root_validator
+    def check_partition_spec(cls, data: Dict[str, Any]):
+        default_spec_id = data["default_spec_id"]
+
+        for spec in data["partition_specs"]:
+            if spec["spec-id"] == default_spec_id:
+                return data
+
+        raise ValueError(f"default-spec-id {default_spec_id} can't be found")
+
+    @root_validator(skip_on_failure=True)
```

Comment:
Why skip on failure? If the order ID is defined, it should match a sort order.

Reply:
What this means is that if there are missing fields, then we don't want to run the validation of the `sort_order`, because we have other problems (maybe missing fields). Therefore I figured to only run the validation if all prior validations have succeeded.

From: https://pydantic-docs.helpmanual.io/usage/validators/#root-validators

As with field validators, root validators can have `pre=True`, in which case they're called before field validation occurs (and are provided with the raw input data), or `pre=False` (the default), in which case they're called after field validation.

Field validation will not occur if `pre=True` root validators raise an error. As with field validators, "post" (i.e. pre=False) root validators by default will be called even if prior validators fail; this behaviour can be changed by setting the `skip_on_failure=True` keyword argument to the validator. The `values` argument will be a dict containing the values which passed field validation and field defaults where applicable.

Reply:
Got it. So just don't run if other validators have failed. Makes sense!